lib/anno_utils.pm

```diff
@@ -0,0 +1,2917 @@
+
+        # According to the genome loading code, the function must be added to the product attribute (go figure)
+        # https://github.com/kbaseapps/GenomeFileUtil/blob/master/lib/GenomeFileUtil/core/FastaGFFToGenome.py#L665-L666
+        # Note that this overwrites anything that was originally in the 'product' field if it previously existed
+        # Also note that I'm forcing every feature to have at least an empty product field
+        if (!defined($ftr_attributes->{'product'})) {
+            $ftr_attributes->{'product'}='';
+        }
+
+        #Look for, and add function
+        if(exists($ftrs_function_lookup{$ftr_attributes->{'id'}})) {
+            $ftr_attributes->{'product'}=$ftrs_function_lookup{$ftr_attributes->{'id'}}->{'functions'};
+        }
+
+        $gff_line->[8] = $ftr_attributes;
+        push(@new_gff_contents, $gff_line);
+    }
+    print "INFO: Updated new_gff_contents has ". scalar @new_gff_contents . " entries\n";
+    return \@new_gff_contents;
+}
+
+sub _write_gff {
+    my ($self, $gff_contents, $gff_filename, $attr_delimiter) = @_;
+
+    # Open $gff_filename to write the @$gff_contents array back to the file
+    my $fh = $self->_openWrite($gff_filename);
+    # Loop over the array
+    foreach (@$gff_contents) {
+        if(scalar(@$_)>=9){
+            # Reform the attributes string
+            my $attributes = $_->[8];
+            my @attributes = ();
+            foreach my $key ( sort keys %$attributes ) {
+                my $attr=$key.$attr_delimiter.$attributes->{$key};
+                push(@attributes,$attr);
+            }
+            $_->[8] = join(";",@attributes);
+        }
+        print $fh join("\t", @$_)."\n";
+    }
+    close $fh;
+    unless (-e $gff_filename) {
+        croak "**In _write_gff ERROR: could not find file $gff_filename\n";
+    }
+    unless (-s $gff_filename) {
+        croak "**In _write_gff ERROR: empty file $gff_filename\n";
+    }
+    print "\n*********First 20 lines of the newly written GFF file:\n";
+    $self->_print_fasta_gff(0, 20, $gff_filename);
+}
+
+
+sub _parse_fasta {
+    my ($self, $fasta_filename) = @_;
+    print "Parsing FASTA contents from file $fasta_filename \n";
+
+    my @fasta_lines = ();
+    # Open $fasta_filename to read into an array
+    my $fh = $self->_openRead($fasta_filename);
+    chomp(@fasta_lines = <$fh>);
+    close($fh);
+
+    # Read in the fasta
+    my %fasta_contents = ();
+    my $seqio = Bio::SeqIO->new( -file => $fasta_filename, -format => 'fasta' );
+    while(my $seq_obj = $seqio->next_seq) {
+        $fasta_contents{$seq_obj->id}=$seq_obj->seq;
+    }
+    return \%fasta_contents;
+}
+
+
+sub _extract_cds_sequences_from_fasta {
+    my ($self, $fasta_contents, $gff_contents) = @_;
+
+    my %gene_seqs = ();
+    foreach my $gff_line (@$gff_contents) {
+        #First, there are gff "lines" that are empty
+        if(scalar(@$gff_line)<9) {
+            next;
+        }
```

Comment:
The `9` here looks like a magic number.

you can also write it as `next if @$gff_line < 9` to be more concise.

Reply:
The magic number 9 happens to be the expected number of segments of each gff line.